src/client/cli/cmd/alias.cpp

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright (C) 2021 Canonical, Ltd.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 3.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "alias.h"
+#include "common_cli.h"
+
+#include <multipass/cli/argparser.h>
+
+namespace mp = multipass;
+namespace cmd = multipass::cmd;
+
+mp::ReturnCode cmd::Alias::run(mp::ArgParser* parser)
+{
+    auto ret = parse_args(parser);
+    if (ret != ParseCode::Ok)
+    {
+        return parser->returnCodeFrom(ret);
+    }
+
+    aliases.add_alias(alias_name, alias_definition);
+
+    // TODO: create symlink/script to directly execute the alias
+
+    return ReturnCode::Ok;
+}
+
+std::string cmd::Alias::name() const
+{
+    return "alias";
+}
+
+QString cmd::Alias::short_help() const
+{
+    return QStringLiteral("Create an alias");
+}
+
+QString cmd::Alias::description() const
+{
+    return QStringLiteral("Create an alias to be executed on a given instance.");
+}
+
+mp::ParseCode cmd::Alias::parse_args(mp::ArgParser* parser)
+{
+    parser->addPositionalArgument("name", "Name given to the alias being defined", "<name>");
+    parser->addPositionalArgument("definition", "Alias definition in the form <instance>:'<command> [<args>]'",
+                                  "<definition>");
+
+    auto status = parser->commandParse(this);
+    if (status != ParseCode::Ok)
+        return status;
+
+    if (parser->positionalArguments().count() != 2)
+    {
+        cerr << "Wrong number of arguments given\n";
+        return ParseCode::CommandLineError;
+    }
```

Comment:
Yeah, totally agree about not splitting args on spaces and I think option 2 is the best choice.

Reply:
Thanks! About option 2: is `[--]` optional? Shouldn't it be `<name> <instance> <command> [-- <args> ...]`?

Reply:
It is optional, you only need it if any of the `<args>` starts with a `-`. All positional arguments beyond `<command>` are arguments to the alias, so there's no need for it.

Reply:
+1 for no spaces in cmd, preferring no colon, and matching `exec` — this problem was already solved for that cmd, which works quite well. The only thing that surprised me is the placement of the `[--]`. Shouldn't it come before `<command>`?

Reply:
> The only thing that surprised me is the placement of the [--]. Shouldn't it come before <command>?

Probably should, yes:
```plain
$ multipass help exec
Usage: multipass exec [options] <name> [--] <command>
…
```

Reply:
The only problem I've with option 2 is that we now have 3 positional arguments. It becomes unwieldy and unclear whether `<name>` comes first, or does `<command>`? The most common use will be just a plain:
```plain
multipass alias cowsay instance cowsay
```

We should probably support a short-hand:
```plain
multipass alias cowsay instance
```

The order is awkward, too (the common "target", "second target" problem of `ln`). I think the most intuitive would be to do (and that's the order `snap` has it:
```plain
multipass alias instance:cowsay cowsay
# shorthand
multipass alias instance:cowsay
```

But then the handling of arguments becomes tricky:
```
# this is likely wrong
multipass alias instance:cowsay arg1 arg2
# as is this
multipass alias instance:cowsay -- arg1 arg2
# these would be the correct ones
multipass alias instance:cowsay cowsay arg1 arg2
multipass alias instance:cowsay -- cowsay arg1 arg2
```

I'm starting to think we should skip arguments altogether for now and see if we get feature requests for that.

Reply:
So I think I'm changing my vote to option 3:

3. `multipass alias <instance>:<command> [<alias>]`

Reply:
And in case we skip arguments and stick to 3, who wants arguments can create a script on the instance. But in case we got requests to add arguments to aliases, I think we'll have to add a new syntax to define an alias with commands, thus complicating the interface.

Reply:
Hmm, not allowing arguments and requiring a script/alias/fn inside the instance would be quite limiting. But the arguments would still be allowed if quoted with option 3, right? So the mismatch with `exec` worries me more.

The awkwardness in `ln` isn't intrinsic, is it? Other commands have the target first (e.g. `tar`). It's often the case when an equals sign is implicit (e.g. `git config --global alias.co checkout`). With `ln`, I think it's rather the mismatch with `cp` and `mv` that makes it awkward. I fear we could be causing the same feel for someone coming from `exec`:

```plain
# if you're used to 
multipass -v exec foo -- ip -br -c a
# it's perhaps less awkward to say
multipass -v alias fooip foo -- ip -br -c -a
# then
multipass -v alias foo:"ip -br -c -a" fooip
# idk
```

Reply:
Let me summarize what we talked about:

- `ssh` (the client, not the protocol) is weird (aka broken) in that:
  ```bash
  $ ssh localhost mkdir foo\ bar
  $ ssh localhost ls
  foo
  bar
  $ ssh localhost echo \$HOME
  /home/ubuntu
  ```
  I.e. it always has a shell interpret the command line. So the user has to be very careful on how to quote things to become separate arguments.
- `multipass exec` is not using a shell:
  ```bash
  $ multipass exec primary mkdir foo\ bar
  $ multipass exec primary ls
  'foo bar'
  $ multipass exec primary echo \$HOME
  $HOME
  ```
  We definitely want to maintain that in the `alias` feature, that exactly the arguments provided are passed through to the `<command>` executable in the instance. If we support arguments, you could still have a shell interpret the command with `sh -c "..."`. If `<command>` includes special characters, the executable has to be called exactly that.
- `alias` from `exec` is different in that `alias` is creating an entity (the alias), and it's prevalent in CLIs that the thing you create is last on the command line:
  ```bash
  $ cp <src> <tgt>
  $ mv <src> [<src> ...] <tgt>
  $ ln <target> [<link>]
  $ snap alias <snap.app> [<alias>]
  ```
- While awkward, it's still possible to allow, with time, the addition of arguments, possibly like so:
  ```plain
  multipass alias <instance>:<command> [<alias>] [-- <arg> …]
  ```
- There will still be even more complex use cases that are probably out of scope, e.g. templating/placeholders.
- The most common use case of aliases will be exporting a command from inside the instance to the host, and so I'd like to optimize for this:
  ```plain
  multipass alias <instance>:<command> [<alias>]
  ```
- Note this still supports arguments when _invoking_ the alias:
  ```plain
  multipass <alias> [<arg> ...]
  <alias> [<arg> ...]
  ```
  Equivalent to:
  ```plain
  multipass exec <instance> -- <command> [<arg> ...]
  ```

I vote for this last incantation for this.

Reply:
Hey, thanks for the comments! I've been considering all this. Now I'm convinced that, in terms of usability and simplicity, `multipass alias <instance>:<command> [<aliasname>]` would be the best choice at the moment. I don't think adding parameters after `--` would be a big issue. In fact, since we have three positional arguments being one of them optional, we can't avoid using `--`. Circumventing the use of `--` would imply to either make one argument mandatory, thus losing the simplicity of the approach.